qcodes/utils/debug.py

```diff
@@ -0,0 +1,124 @@
+import socket
+import time
+import re, sys, types, inspect
+
+DEFAULT_LOG = sys.stdout
+
+
+def formatAllArgs(args, kwds):
+    """
+    makes a nice string representation of all the arguments
+    """
+    allargs = []
+    for item in args:
+        allargs.append(str(item))
+    for key, item in kwds.items():
+        allargs.append(f"{key}={item}")
+    formattedArgs = ", ".join(allargs)
+    if len(formattedArgs) > 150:
+        return formattedArgs[:146] + " ..."
+    return formattedArgs
+
+
+def logfunction(theFunction, log, displayName=None):
+    """Decorator to log whenever a function is called including arguments"""
+    if not displayName:
+        displayName = theFunction.__name__
+
+    def _wrapper(*args, **kwds):
+        argstr = formatAllArgs(args, kwds)
+
+        # Log the entry into the function
+        if isinstance(log, socket.socket):
+            log.send(f"{displayName}({argstr})\n".encode())
+        else:
+            print(f"{displayName}({argstr})", file=log)
+            log.flush()
+            time.sleep(2e-3)
+
+        # t0 = time.perf_counter()
+        returnval = theFunction(*args, **kwds)
+        # print(f" | {time.perf_counter() - t0:.3g} s", file=log)
+        # log.flush()
+
+        # Log return
+        ##indentlog("return: %s"% str(returnval)
+        return returnval
+
+    return _wrapper
+
+
+def logmethod(theMethod, log, displayName=None):
+    """Decorator to log whenever a method is called including arguments"""
+
+    def _methodWrapper(self, *args, **kwds):
+        "Use this one for instance or class methods"
+
+        argstr = formatAllArgs(args, kwds)
+        if isinstance(log, socket.socket):
+            log.send(f"{displayName}.{theMethod.__name__}({argstr})\n".encode())
+        else:
+            print(f"{displayName}.{theMethod.__name__}({argstr})", file=log)
+            log.flush()
+            time.sleep(2e-3)
+
+        # t0 = time.perf_counter()
+        returnval = theMethod(self, *args, **kwds)
+        # print(f" | {time.perf_counter() - t0:.3g} s", file=log)
+        # log.flush()
+
+        return returnval
+
+    return _methodWrapper
+
+
+def logclass(
+    cls, methodsAsFunctions=False, log=None, logMatch=".*", logNotMatch="asdfnomatch"
```

Comment:
Things start getting a bit cryptic from this point. What is the purpose of logMatch and logNotMatch? Why is logNotMatch defaulting to a random string `asdf`?

Reply:
Better?

Reply:
Yeah, it's much clearer. I did a quick google for unmatchable regex strings and it came back with /(?!)/

I just looked into a 'single regex' construction for doing positive and negative matching (i.e. the union of positive matches and NOT the negative matches) and while it's not horribly complicated I agree that this is just easier.

I have 2 suggestions:
- Replace `asdfnomatch` with `(?!)` and mention it in the docstring
- Perhaps allow log_match/log_not_match to be given as a list of strings for non-regexperts, and then combine them à la
```
log_match = ['banana', 'berry', 'apple']
log_match_re = '(banana)|(berry)|(apple)'
```

Reply:
Like this?